### DISCUSSION TITLE
How to set default workspace / project running via Helm chart

### DISCUSSION THREAD

**Discussion by nodesocket** [coder/code-server#3081] [Q&A]
I am running the official Helm chart, so Kubernetes and Docker. For the life of me, I can't figure out how to change the default project / workspace. I navigate to `File => Open` select the directory I want to be the default and close the browser tab. When I reopen the browser it's still defaulting into `/home/coder`.

It seems running via Docker and Helm chart, that code-server always defaults to `/home/coder` no matter what the last project was or last open path was.

How can I modify to always start in `/home/coder/code` instead?

**Comment by code-asher** [comment]
Interesting, it's supposed to store the last visited in `~/.local/share/code-server/coder.json` and read that every time the page is opened.

I'm not familiar with the Helm charts but the ways I can think it wouldn't be used are if:

1. ?file or ?folder was in the URL (these override the last visited)
2. A path argument was provided to code-server on launch (for example `code-server /home/coder`) (also overrides last visited)
3. ?ew was in the URL (clears the last visited)

Do any of those seem to apply in your situation? If not passing `/home/coder/code` to code-server as an argument should do the trick although at the cost of disabling the last visited functionality.

**Reply by nodesocket** [reply]
Thanks for the reply.

`~/.local/share/code-server/coder.json` indeed exists and has the contents of:

```
{
  "query": {},
  "update": {
    "checked": 1618261174656,
    "version": "3.9.3"
  },
  "lastVisited": {
    "url": "/home/coder",
    "workspace": false
  }
}
```

However, even when I close the default workspace `file => close folder`, click `file => open` and select my desired default it does not updated the `coder.json` file. Then when I navigate back, it is still using `/home/coder`

**Reply by nodesocket** [reply]
Correction, it does update `coder.json`, however when I close the browser tab and reload, it reverts `url` back to `/home/coder`. I verified in my Helm chart I am not setting a default path argument.

**Reply by code-asher** [reply]
Weird! Here's what I tried to do to replicate locally without any luck:

1. Open `localhost:8080`
2. Verify I'm at `/home/coder`
3. `mkdir code`
3. Use `File` > `Open Folder` and browse to `/home/coder/code`
3. Redirects to `localhost:8080?folder=/home/coder/code`
4. Close tab once loaded
5. Open a new tab and go to `localhost:8080`
6. Verify I'm at `/home/coder/code`

So perhaps it's something specific to our Helm chart but I'm not sure what it could be.

**Reply by nodesocket** [reply]
@code-asher I made a quick video showing my exact steps. Perhaps it will helps us debug this?


https://user-images.githubusercontent.com/523312/114476226-7772c080-9bbf-11eb-967d-f1a0a0f4deed.mp4

**Reply by code-asher** [reply]
Very helpful! This is bizarre. 

The only difference I see immediately is that I have "Open Folder" instead of "Open". I _think_ one shows on Linux/Windows and the other shows on Mac but it seems very unlikely the client would be having an effect here. Just in case @jsjoeio does this repro for you?

Perhaps this needs Kubernetes to repro although that seems odd.

The only other thing that immediately comes to mind is if a directory is specified in the config file (`~/.config/code-server/config.yaml`) but I don't think you even can specify the directory in there (at least, it's not working for me).

**Reply by jsjoeio** [reply]
This is super strange. So running locally on my mac, I get the same results as you @code-asher

## Video - macOS, running code-server locally

https://user-images.githubusercontent.com/3806031/114794000-6f8e5a00-9d40-11eb-9534-c20ace5130b5.mov

BUT when I use Railway, I'm seeing the same behavior as @nodesocket 🤔 

## Video - code-server running with Railway

[Link to setup](https://github.com/jsjoeio/code-server-railway2)

https://user-images.githubusercontent.com/3806031/114794133-ab292400-9d40-11eb-99a4-92e56107da21.mov

**Reply by nodesocket** [reply]
@jsjoeio What is Railway? I am just using the official Helm chart.

**Reply by jsjoeio** [reply]
It's a platform for running/hosting applications like code-server. We have an official guide [here](https://github.com/cdr/deploy-code-server/blob/main/guides/railway.md)

[Link to Railway](https://railway.app/)

**Reply by nodesocket** [reply]
@jsjoeio here is another detailed video I did showing everything again _(turn on sound)_. Hopefully it's helpful.


https://user-images.githubusercontent.com/523312/114800557-8d1eec00-9d5f-11eb-9dd4-2fe43ce30ff6.mp4

**Reply by code-asher** [reply]
Interesting, it replicates for me in Railway as well. Hopefully it's analogous to whatever problem is going on with the Helm chart.

Maybe there is something in these deployments that writes to `coder.json` when a client connects?

**Reply by jsjoeio** [reply]
That was helpful! Thanks for doing that again. So I guess what happens in your environment (and my Railway):
1. It updates `coder.json` as expected
2. When you open a new tab, it overwrites `coder.json` and then loads the default home

I don't know why this is happening on some environments but not others 🤔

I tried reproducing locally again using `--link` to see if it made a difference, but nothing.  

https://user-images.githubusercontent.com/3806031/114915875-04449680-9dd9-11eb-85af-67b7501f1f72.mov

Bizarre. @code-asher any thoughts on what might be causing this?

**Reply by code-asher** [reply]
OK in the Railway case there is a directory being passed to code-server which overrides the last visited behavior: https://github.com/code-asher/deploy-code-server/blob/main/deploy-container/entrypoint.sh#L81

Maybe the same is happening with the Helm chart but I haven't seen anything like that and @nodesocket also mentioned above there is no path argument being set.

**Reply by jsjoeio** [reply]
> OK in the Railway case there is a directory being passed to code-server which overrides the last visited behavior:

I feel like @bpmct would want to know about that. Tagging just so you know since you've worked with the Railway folks directly.

**Reply by code-asher** [reply]
I deployed the Helm chart locally via Kind and was able to reproduce so hopefully I can track this down.

**Reply by code-asher** [reply]
Ahh so I think it pulls our Docker image. Looks like it's this (notice the `.`):

https://github.com/cdr/code-server/blob/97fbbfaecc5975ed3ba18e6bef9b4ee6c40ac0ac/ci/release-image/Dockerfile#L46

**Reply by code-asher** [reply]
So I suppose we need to make this configurable rather than hard-coding to `.`.
